src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7164,12 +7164,54 @@ bool canFollowNullableType()
                     case ParseTypeMode.AfterIs:
                     case ParseTypeMode.DefinitePattern:
                     case ParseTypeMode.AsExpression:
-                        // These contexts might be a type that is at the end of an expression. In these contexts we only
-                        // permit the nullable qualifier if it is followed by a token that could not start an
-                        // expression, because for backward compatibility we want to consider a `?` token as part of the
-                        // `?:` operator if possible.
-                        //
-                        // Similarly, if we have `T?[]` we do want to treat that as an array of nullables (following
+                        // These contexts might be a type that is at the end of an expression.
+                        // For backward compatibility we want to consider a `?` token as part of the `?:` operator if possible.
+                        // However, if current token is an identifier, it might be beneficial to allow `?`
+                        // and treat this identifier as a designation. Nullable types in patterns are semantically invalid,
+                        // so we will get a nice error about that during binding
```

Comment:
In general, i think giving examples is good.

Reply:
Added explanation comment with examples for every token/token type in the list below. Is there anything else, which needs clarification?